src/Microsoft.ML.PipelineInference/AutoMlUtils.cs

```diff
@@ -15,21 +15,38 @@ namespace Microsoft.ML.Runtime.PipelineInference
 {
     public static class AutoMlUtils
     {
-        public static AutoInference.RunSummary ExtractRunSummary(IHostEnvironment env, IDataView data, string metricColumnName)
+        public static AutoInference.RunSummary ExtractRunSummary(IHostEnvironment env, IDataView result, string metricColumnName, IDataView trainResult = null)
         {
-            double metricValue = 0;
+            double testingMetricValue = 0;
+            double trainingMetricValue = -1d;
             int numRows = 0;
-            var schema = data.Schema;
-            schema.TryGetColumnIndex(metricColumnName, out var metricCol);
+            var schema = result.Schema;
+            bool hasIndex = schema.TryGetColumnIndex(metricColumnName, out var metricCol);
+            env.Check(hasIndex);
 
-            using (var cursor = data.GetRowCursor(col => col == metricCol))
+            using (var cursor = result.GetRowCursor(col => col == metricCol))
             {
                 var getter = cursor.GetGetter<double>(metricCol);
-                cursor.MoveNext();
-                getter(ref metricValue);
+                bool moved = cursor.MoveNext();
+                env.Check(moved);
+                getter(ref testingMetricValue);
             }
 
-            return new AutoInference.RunSummary(metricValue, numRows, 0);
+            if (trainResult != null)
+            {
+                var trainSchema = trainResult.Schema;
+                env.Check(trainSchema.TryGetColumnIndex(metricColumnName, out var trainingMetricCol));
+
+                using (var cursor = trainResult.GetRowCursor(col => col == trainingMetricCol))
+                {
+                    var getter = cursor.GetGetter<double>(trainingMetricCol);
+                    bool moved = cursor.MoveNext();
+                    env.Check(moved);
```

Comment:
>env.Check(moved); [](start = 20, length = 17)

We might want to make this error exception a little more specific, since the default message here, "The operation failed due to the current state of the object" (IIRC) is, while not strictly incorrect, perhaps not helpful from the perspective of allowing the user to diagnose the problem.

Reply:
What error message would you suggest?

Reply:
HI George, so I think the idea is that you're expecting that the the input `trainResult`, if present, will have exactly one row. (Indeed, this seems to be the case in many of these methods... is it possible to just use an `IRow` rather than deal with an `IDataView`? We sometimes use an `IRow` for such purposes, rather than deal with the bother of having to open cursors, etc.

So I'd suggest something along those lines, than what I see currently... "could not move cursor forward" is certainly describing what's going on sort of, but the real error isn't that it can't move the cursor, or in the second case that you *can*, but rather that you expected an `IDataView` with exactly one row.

---
In reply to: [188125581](https://github.com/dotnet/machinelearning/pull/152#discussion_r188125581) [](ancestors = 188125581)

Reply:
You're doing this a *lot* in this and other similar functions I see.

---
In reply to: [189667233](https://github.com/dotnet/machinelearning/pull/152#discussion_r189667233) [](ancestors = 189667233,188125581)

Reply:
Made some changes to this, following your suggestion to extract the code to a separate utility function. I'll push the changes shortly.